app/utils/context.tsx

```diff
@@ -0,0 +1,100 @@
+"use client";
+import React, {
+  createContext,
+  useContext,
+  useState,
+  useEffect,
+  ReactNode,
+} from "react";
+
+interface WalletState {
+  provider: any;
+  address: string | null;
+  network: string | null;
+}
+
+interface AppStateContextType {
+  isWalletModalOpen: boolean;
+  toggleWalletModal: () => void;
+  wallet: WalletState | null;
+  setWalletState: (wallet: WalletState) => void;
+}
+
+const AppStateContext = createContext<AppStateContextType | undefined>(
+  undefined
+);
+
+export const AppStateProvider = ({ children }: { children: ReactNode }) => {
+  const [isWalletModalOpen, setIsWalletModalOpen] = useState(false);
+  const [wallet, setWalletState] = useState<WalletState | null>(null);
+
+  const toggleWalletModal = () => {
+    setIsWalletModalOpen((prevState) => !prevState);
+  };
+
+  useEffect(() => {
+    const provider = wallet?.provider;
+
+    if (provider) {
```

Comment:
i prefer the golang-style invariant handling where indentation implies a off-nominal condition. obviously not applicable in all cases but i think it would be more clear

```
if (!provider) {
        console.warn(
          "Provider does not support event listeners. Consider adding polling for updates."
        );
        return;
}